lib/crawler/engine.rb

```diff
@@ -55,6 +70,25 @@ def extract_links(url:)
       self
     end
 
+    def report_save(path: '')
+      save_path = path.empty? ? REPORT_SAVE_PATH : path
+      File.write(save_path, @report.to_yaml)
+    end
+
+    def before_crawling
+      sign_in if ENV['username']
+      sleep 5
+    end
+
+    def after_crawling
+    end
+
+    def overwrite_callback(method:, &block)
```

Comment:
Good for now. In the future we might want to define an array of possible callbacks, so that there is no need to overwrite.
Just like in active record: `before_save` adds another callback to the list, `after_save` does the same.  `overwrite_callback` sounds too technical and require developer to know what to overwrite anyway.